### PR TITLE
coprocessor/endpoint: check range for dag and analyze

### PR DESCRIPTION
**Pull request by AndreMouche**
For coprocessor request with type is dag or analyze, we check req's ranges to make sure the scanned keys should always be table record or index. @BusyJay @hicqu PTAL

### CODE REVIEW

**Comment by AndreMouche** [comment]
address comments @BusyJay

File: src/coprocessor/dag/executor/index_scan.rs [hunk 1]
```diff
@@ -353,7 +355,8 @@ mod test {
         let (snapshot, start_ts) = wrapper.store.get_snapshot();
         let store = SnapshotStore::new(snapshot, start_ts, IsolationLevel::SI, true);
 
-        let mut scanner = IndexScanExecutor::new(wrapper.scan, wrapper.ranges, store, false);
+        let mut scanner =
+            IndexScanExecutor::new(wrapper.scan, wrapper.ranges, store, false).unwrap();
```
Comment (ngaut) [inline src/coprocessor/dag/executor/index_scan.rs]: Usually, we should not use unwrap method for production env. Please log error message and return error.

Comment (AndreMouche) [reply src/coprocessor/dag/executor/index_scan.rs]: It's only used in the unit tests @ngaut

**Comment by AndreMouche** [comment]
@BusyJay @hicqu PTAL

File: src/coprocessor/codec/table.rs [hunk 2]
```diff
@@ -65,6 +65,15 @@ pub fn extract_table_prefix(key: &[u8]) -> Result<&[u8]> {
     }
 }
 
+/// Check if the range is for table record or index.
+pub fn check_table_ranges(range: &[KeyRange]) -> Result<()> {
+    for range in range.iter() {
```
Comment (BusyJay) [inline src/coprocessor/codec/table.rs]: I don't think `iter()` is necessary.

File: tests/coprocessor/test_analyze.rs [hunk 3]
```diff
@@ -206,3 +206,23 @@ fn test_analyze_index() {
...
+
+    let product = ProductTable::new();
+    let (_, end_point) = init_data_with_commit(&product, &data, true);
+    let mut req = new_analyze_index_req(&product.table, 3, product.name.index, 4, 32);
+    let mut key_range = KeyRange::new();
+    key_range.set_start(b"xxx".to_vec());
+    key_range.set_end(b"zzz".to_vec());
+    req.set_ranges(RepeatedField::from_vec(vec![key_range]));
+    let resp = handle_request(&end_point, req);
+    assert!(!resp.get_other_error().is_empty());
```
Comment (BusyJay) [inline tests/coprocessor/test_analyze.rs]: Should stop the `end_point`.

Comment (AndreMouche) [reply tests/coprocessor/test_analyze.rs]: address comments

**Comment by AndreMouche** [comment]
@BusyJay @hicqu PTAL

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**